src/eei.cpp

```diff
@@ -492,18 +492,25 @@ namespace hera {
       evmc_result call_result;
 
       // Start with base call gas
-      int64_t extra_gas = GasSchedule::call;
+      takeInterfaceGas(GasSchedule::call);
 
+      // These checks are in EIP150 but not in the YellowPaper
       // Charge valuetransfer gas if value is being transferred.
-      // Only charge callNewAccount gas if the account is new and value is being transferred per EIP161.
-      if (!isZeroUint128(call_message.value)) {
-        extra_gas += GasSchedule::valuetransfer;
+      if ((kind == EEICallKind::Call || kind == EEICallKind::CallCode) && !isZeroUint128(call_message.value)) {
+        takeInterfaceGas(GasSchedule::valuetransfer);
+
+        // Only charge callNewAccount gas if the account is new and value is being transferred per EIP161.
         if ((kind == EEICallKind::Call) && !m_context->fn_table->account_exists(m_context, &call_message.destination))
-          extra_gas += GasSchedule::callNewAccount;
+          takeInterfaceGas(GasSchedule::callNewAccount);
       }
 
-      // This check is in EIP150 but not in the YellowPaper
-      takeInterfaceGas(extra_gas);
+      if (m_msg.depth >= 1024)
```

Comment:
This is weird (although correct) that the depth is checked _after_ e.g. new account creation cost is already charged.

Reply:
And in EVM geth the depth is checked before the account is charged; in its current state, wagon relies a lot on the default code and it would misbehave and force me to write a lot of redundant code.

Reply:
Well, I'm confused every time I try to understand how calls works. 

First of all, the order does not matter. All those checks are applied to the caller so if any of these fails the caller execution will be aborted.

I'm for moving the depth check up front because it's the cheapest one.
My ideal order would be:

1. depth check (no state access)
2. value transfer additional cost check (no state access)
3. balance check (current account access)
4. account creation additional cost check (external account lookup)


Side note: the order will matter if we decide to always user REVERT semantic (i.e. dropping OOG behavior).